### PR TITLE
FullDebugMode v1.65

### PR DESCRIPTION
**Pull request by TetzkatLipHoka**

### CODE REVIEW

**Review by maximmasiutin — CHANGES_REQUESTED** [review]

File: FullDebugMode DLL/Resource/ZLib/FastMM_FullDebugMode.res
[the remark this answers is not in the document]
Comment (TetzkatLipHoka) [reply FullDebugMode DLL/Resource/ZLib/FastMM_FullDebugMode.res:1]: Those resource files contain the DLL in compressed form to embed them into the Application

File: FullDebugMode DLL/Precompiled/madExcept/FastMM_FullDebugMode.dll
[the remark this answers is not in the document]
Comment (TetzkatLipHoka) [reply FullDebugMode DLL/Precompiled/madExcept/FastMM_FullDebugMode.dll:1]: Generally you can build them from IDE (so do I), I can provide you however with a commandline example on how to build them from prompt if you need.
Note the specific builds require either madExcept, Eureka or Jedi VCL as they are used for the stackwalk/symbol-resolve and hence need to be linked in the commandline.

File: FullDebugMode DLL/FastMM_FullDebugMode.dpr [hunk 1]
```diff
@@ -830,6 +843,15 @@ begin
   LLocationCacheInitialized := False;
 
   Result := ABuffer;
+
+  {$IF Declared( AtomicCmpExchange )}
+  {This routine is protected by a lock - only one thread can be inside it at any given time.}
+  while AtomicCmpExchange(LLogStackTrace_Locked, 1, 0) <> 0 do
```
[the remark this answers is not in the document]
Comment (TetzkatLipHoka) [reply FullDebugMode DLL/FastMM_FullDebugMode.dpr:849]: Yes, can do that in next update if you wish. So far I kept it as close to the original version from FastMM4/5 as possible for easier maintance. Changes made were only to build on older IDEs for smaller DLLs

File: FastMM4.pas [hunk 2]
```diff
@@ -20401,9 +20401,9 @@ procedure FinalizeMemoryManager;
       ReportLockContention;
 {$ENDIF}
 {$IFNDEF NeverUninstall}
-      {$IF Defined( FullDebugMode ) AND Defined( LoadDebugDLLDynamically )}
```
[the remark this answers is not in the document]
Comment (TetzkatLipHoka) [reply FastMM4.pas:20404]: Thanks for pointing that out. Probably commented it during testing and forgot about it. Will fix that.

**Comment by maximmasiutin** [comment]
Thank you for the explanation! I resolved the conflicts with the recent origin changes. Can you please check whether the final version is OK? Otherwise, can you do a new pull request to fix the leftovers?

**Comment by TetzkatLipHoka** [comment]
Sure, was just about to fix the "issues" from this convo and pull your latest changes for rebuilding :)
Gimme some minutes

**Pull request merged**